crypto/schnorr.c

```diff
@@ -0,0 +1,254 @@
+/**
+ * Copyright (c) 2021 The Bitcoin ABC developers
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "schnorr.h"
+#include "hmac_drbg.h"
+#include "memzero.h"
+#include "rfc6979.h"
+
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+
+int jacobi(const bignum256 *_n, const bignum256 *_k) {
+  int t = 0;
+  int k_is_one;
+  uint32_t r;
+  bignum256 n, k, temp;
+
+  assert(!bn_is_zero(_k) && bn_is_odd(_k));
+
+  bn_copy(_n, &n);
+  bn_copy(_k, &k);
+
+  bn_mod(&n, &k);
+
+  while (!bn_is_zero(&n)) {
+    while (bn_is_even(&n)) {
+      bn_rshift(&n);
+      r = k.val[0] & 0x07;
+      t ^= (r == 3 || r == 5);
+    }
+
+    bn_cmov(&temp, 0, &n, &n);
+    bn_cmov(&n, 0, &k, &k);
+    bn_cmov(&k, 0, &temp, &temp);
```

Comment:
Why don't you use `bn_copy` instead of `bn_cmov`?

Reply:
Done in bc6267c